components/script/dom/bindings/utils.rs

```diff
@@ -178,28 +181,46 @@ pub unsafe fn get_property_on_prototype(
 
 /// Get an array index from the given `jsid`. Returns `None` if the given
 /// `jsid` is not an integer.
-pub fn get_array_index_from_id(_cx: *mut JSContext, id: HandleId) -> Option<u32> {
+pub unsafe fn get_array_index_from_id(cx: *mut JSContext, id: HandleId) -> Option<u32> {
     let raw_id = id.into();
-    unsafe {
-        if RUST_JSID_IS_INT(raw_id) {
-            return Some(RUST_JSID_TO_INT(raw_id) as u32);
-        }
-        None
+    if RUST_JSID_IS_INT(raw_id) {
+        return Some(RUST_JSID_TO_INT(raw_id) as u32);
     }
-    // if id is length atom, -1, otherwise
-    // return if JSID_IS_ATOM(id) {
-    //     let atom = JSID_TO_ATOM(id);
-    //     //let s = *GetAtomChars(id);
-    //     if s > 'a' && s < 'z' {
-    //         return -1;
-    //     }
-    //
-    //     let i = 0;
-    //     let str = AtomToLinearString(JSID_TO_ATOM(id));
-    //     return if StringIsArray(str, &mut i) != 0 { i } else { -1 }
-    // } else {
-    //     IdToInt32(cx, id);
+
+    // if !RUST_JSID_IS_VOID(raw_id) {
+    //     return None;
     // }
+
+    if !RUST_JSID_IS_STRING(raw_id) {
+        return None;
+    }
+
+    let s = jsstr_to_string(cx, RUST_JSID_TO_STRING(raw_id));
+    if s.len() == 0 {
+        return None;
+    }
+
+    let first = s.chars().next().unwrap();
+    if first.is_ascii_lowercase() {
+        return None;
+    }
+
+    let mut i: u32 = 0;
+    let is_array = if s.is_ascii() {
+        let chars = s.as_bytes();
+        StringIsArrayIndex1(chars.as_ptr() as *const i8, chars.len() as u32, &mut i)
```

Comment:
c_char is a different type on ARM, so we need `as *const _` instead of hardcoding i8.